src/finiteVolume/processes/les.cpp

```diff
@@ -0,0 +1,263 @@
+    CHKERRQ(ierr);
+    ierr = CompressibleFlowComputelesStressTensor(dim, fg, gradAux + aOff_x[VEL], uOff, field, ctx, lesTau);
+    CHKERRQ(ierr);
+
+    // energy equation
+    for (PetscInt ev = 0; ev < flowParameters->numberEV; ++ev) {
+        flux[ev] = 0.0;
+
+        for (PetscInt d = 0; d < dim; ++d) {
+            PetscReal lesevFlux = 0.0;
+            for (PetscInt c = 0; c < dim; ++c) {
+                //
+                lesevFlux += areaMag * sqrt(areaMag) * density * lesTau[d * dim + c] * lesTau[d * dim + c] / mut;
+            }
+
+            //  LESevFlux( rho Di dEVi/dx + rho Di dEVi/dy + rho Di dEVi//dz) . n A +  LESevFlux(-rho ce EV^3/2 ) . n A
+            const int offset = aOff_x[EV_FIELD] + (ev * dim) + d;
+            lesevFlux += -fg->normal[d] * density * (mut * gradAux[offset] - c_e * field[uOff[EV_FIELD] + ev] * sqrt(field[uOff[EV_FIELD] + ev]));
+
+            flux[ev] += lesevFlux;
+        }
+    }
+
+    PetscFunctionReturn(0);
+}
+
+PetscErrorCode ablate::finiteVolume::processes::LES::CompressibleFlowComputelesStressTensor(PetscInt dim, const PetscFVFaceGeom* fg, const PetscReal* gradVel, const PetscInt uOff[],
+                                                                                            const PetscScalar field[], void* ctx, PetscReal* lesTau) {
+    PetscFunctionBeginUser;
+    // pre-compute the div of the velocity field
+    PetscReal divVel = 0.0;
+    PetscReal mut;
+
+    PetscErrorCode ierr;
+    ierr = lesViscosity(dim, fg, field, uOff, ctx, mut);
+    CHKERRQ(ierr);
+
+    for (PetscInt c = 0; c < dim; ++c) {
+        divVel += gradVel[c * dim + c];
+    }
+
+    // March over each velocity component, u, v, w
+    for (PetscInt c = 0; c < dim; ++c) {
+        // March over each physical coordinates
+        for (PetscInt d = 0; d < dim; ++d) {
+            if (d == c) {
+                // for the xx, yy, zz, components
+                lesTau[c * dim + d] = 2.0 * mut * ((gradVel[c * dim + d]) - divVel / 3.0);
+            } else {
+                // for xy, xz, etc
+                lesTau[c * dim + d] = mut * ((gradVel[c * dim + d]) + (gradVel[d * dim + c]));
+            }
+        }
+    }
+    PetscFunctionReturn(0);
+}
+
+PetscErrorCode ablate::finiteVolume::processes::LES::lesViscosity(PetscInt dim, const PetscFVFaceGeom* fg, const PetscScalar field[], const PetscInt uOff[], void* ctx, PetscReal& mut) {
+    PetscFunctionBeginUser;
+    const int euler = 0;
+    const int EV_FIELD = 0;
+
+    auto flowParameters = (DiffusionData*)ctx;
+
+    const PetscReal areaMag = utilities::MathUtilities::MagVector(dim, fg->normal);
+
+    // get the current density from euler
+    const PetscReal density = field[uOff[euler] + CompressibleFlowFields::RHO];
+
+    // get the current ev from ev_field
+    for (PetscInt ev = 0; ev < flowParameters->numberEV; ++ev) {
+        // compute turbulent kinetic energy
+        const PetscReal tke = field[uOff[EV_FIELD] + ev];
```

Comment:
I think that this field is ev*density.  Why are we marching over every ev?  Should this ok be k?